File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.js
function solution(gems) {
    var answer = [];
    
    const set = new Set(gems);
    
    let allGems = set.size;
    
    const map = new Map();
    
    map.set(gems[0], 1);
    
    let left = 0;
    let right = 0;
    
    answer[0] = 0;
    answer[1] = gems.length;
    
    while (left < gems.length && right < gems.length) {        
        if (map.size === allGems) {
            // 최소 거리 체크
            if (answer[1] - answer[0] > right - left) {
                answer[0] = left+1;
                answer[1] = right+1;
            }
            
            map.set(gems[left], map.get(gems[left]) - 1);
            
            if (map.get(gems[left]) === 0) {
                map.delete(gems[left]);
            } 
            
            left++;
        } else {
            right++;
            
            if (right === gems.length) break;
            
            if (!map.has(gems[right])) {
                map.set(gems[right], 1);
            } else {
                map.set(gems[right], map.get(gems[right])+1);
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;

class Solution {
    
    static int min;
    static int[] answer;
    static Set<String> set;
    static Map<String, Integer> map;
    
    public int[] solution(String[] gems) {
        answer = new int[2];
        
        set = new HashSet<>();
        map = new HashMap<>();
        
        int idx = 0;
        
        for (String gem : gems) {
            if (!set.contains(gem)) {
                set.add(gem);
                map.put(gem, idx++);
            }
        }
        
        solve(gems);
        
        return answer;
    }
    
    static void solve(String[] gems) {
        int left = 0;
        int right = 0;
        
        answer[0] = 0;
        answer[1] = gems.length;
        
        Set<String> tmpSet = new HashSet<>();
        int[] arr = new int[set.size()];
        
        tmpSet.add(gems[left]);
        arr[map.get(gems[left])]++;
        
        while (left < gems.length && right < gems.length) {            
            if (tmpSet.size() == set.size()) {
                if (answer[1] - answer[0] > right - left) {
                    answer[0] = left+1;
                    answer[1] = right+1;
                }
                
                arr[map.get(gems[left])]--;
                if (arr[map.get(gems[left])] == 0) tmpSet.remove(gems[left]);
                left++;
            } else {
                right++;
                if (right == gems.length) break;
                tmpSet.add(gems[right]);
                arr[map.get(gems[right])]++;
            }
        }
    }
}    